From 17d1977539d72c315cc81de037b73d6f64ba5367 Mon Sep 17 00:00:00 2001
From: Ilan Peer <ilan.peer@intel.com>
Date: Fri, 22 Dec 2023 11:04:17 +0200
Subject: [PATCH] WLANRTSYS-82700 AP: Update the HE regulatory information AP
 types for the 6 GHz band

Update the HE regulatory information AP types based on IEEE
P802.11-REVme/D4.0. Set the default AP type to VLP. Check for valid
values when setting 'he_6ghz_reg_pwr_type' in the interface
configuration.

Signed-off-by: Ilan Peer <ilan.peer@intel.com>
---
 hostapd/config_file.c        |  6 ++++++
 hostapd/hostapd.conf         |  9 +++++++--
 src/ap/ap_config.c           |  1 +
 src/ap/beacon.c              |  7 ++++---
 src/ap/ieee802_11.c          |  3 ++-
 src/ap/ieee802_11_he.c       |  9 +++------
 src/common/ieee802_11_defs.h | 17 ++++++++++++-----
 7 files changed, 35 insertions(+), 17 deletions(-)

diff --git a/hostapd/config_file.c b/hostapd/config_file.c
index 4c12d7f78..bb0a8ac9c 100644
--- a/hostapd/config_file.c
+++ b/hostapd/config_file.c
@@ -4444,6 +4444,12 @@ static int hostapd_config_fill(struct hostapd_config *conf,
 		}
 	} else if (os_strcmp(buf, "he_6ghz_reg_pwr_type") == 0) {
 		conf->he_6ghz_reg_pwr_type = atoi(pos);
+		if (conf->he_6ghz_reg_pwr_type > HE_REG_INFO_6GHZ_AP_TYPE_MAX) {
+			wpa_printf(MSG_ERROR,
+				   "Line %d: invalid he_6ghz_reg_pwr_type value",
+				   line);
+			return 1;
+		}
 	// ToDo: remove this and keep opensource configuration for SRPS IE
 	} else if (os_strcmp(buf, "he_spatial_reuse_ie_present_in_beacon") == 0) {
 		conf->he_spatial_reuse_ie_present_in_beacon = atoi(pos);
diff --git a/hostapd/hostapd.conf b/hostapd/hostapd.conf
index 23851af94..0b65b0af2 100644
--- a/hostapd/hostapd.conf
+++ b/hostapd/hostapd.conf
@@ -1040,9 +1040,14 @@ wmm_ac_vo_acm=0
 
 # 6 GHz Access Point type
 # This config is to set the 6 GHz Access Point type. Possible options are:
-# 0 = Indoor AP (default)
-# 1 = Standard Power AP
+# 0 = Indoor AP
+# 1 = Standard power AP
+# 2 = Very low power AP (default)
+# 3 = Indoor enabled AP
+# 4 = Indoor standard power AP
 # This has no impact for operation on other bands.
+# See IEEE P802.11-REVme/D4.0, Table E-12 (Regulatory Info subfield encoding)
+# for more details.
 #he_6ghz_reg_pwr_type=0
 
 # Unsolicited broadcast Probe Response transmission settings
diff --git a/src/ap/ap_config.c b/src/ap/ap_config.c
index e0101374e..4e987aa01 100644
--- a/src/ap/ap_config.c
+++ b/src/ap/ap_config.c
@@ -355,6 +355,7 @@ struct hostapd_config * hostapd_config_defaults(void)
 	conf->he_6ghz_max_ampdu_len_exp = 7;
 	conf->he_6ghz_rx_ant_pat = 1;
 	conf->he_6ghz_tx_ant_pat = 1;
+	conf->he_6ghz_reg_pwr_type = HE_REG_INFO_6GHZ_AP_TYPE_VLP;
 	conf->he_oper_max_co_hosted_bssid_ind = 5;
 	conf->he_oper_cohosted_bss = 1;
 	conf->bss_color_change_timeout  = BSS_COLOR_CHANGE_TIMEOUT_DEFAULT;
diff --git a/src/ap/beacon.c b/src/ap/beacon.c
index 5244cdb9a..233a57bb1 100644
--- a/src/ap/beacon.c
+++ b/src/ap/beacon.c
@@ -673,7 +673,7 @@ static size_t hostapd_probe_resp_elems_len(struct hostapd_data *hapd,
 			 /* An additional Transmit Power Envelope element for
 			  * subordinate client */
 			if (hapd->iconf->he_6ghz_reg_pwr_type ==
-			    HE_6GHZ_INDOOR_AP)
+			    HE_REG_INFO_6GHZ_AP_TYPE_INDOOR)
 				buflen += 4;
 		}
 	}
@@ -2183,7 +2183,8 @@ static u8 * hostapd_gen_fils_discovery(struct hostapd_data *hapd, size_t *len)
 	/* Transmit Power Envelope element(s) */
 	if (is_6ghz_op_class(hapd->iconf->op_class)) {
 		total_len += 4;
-		if (hapd->iconf->he_6ghz_reg_pwr_type == HE_6GHZ_INDOOR_AP)
+		if (hapd->iconf->he_6ghz_reg_pwr_type ==
+		    HE_REG_INFO_6GHZ_AP_TYPE_INDOOR)
 			total_len += 4;
 	}
 #endif /* CONFIG_IEEE80211AX */
@@ -2352,7 +2353,7 @@ int ieee802_11_build_ap_params(struct hostapd_data *hapd,
 			 /* An additional Transmit Power Envelope element for
 			  * subordinate client */
 			if (hapd->iconf->he_6ghz_reg_pwr_type ==
-			    HE_6GHZ_INDOOR_AP)
+			    HE_REG_INFO_6GHZ_AP_TYPE_INDOOR)
 				tail_len += 4;
 		}
 	}
diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index 25673d1cb..a787ec843 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -7721,7 +7721,8 @@ u8 * hostapd_eid_txpower_envelope(struct hostapd_data *hapd, u8 *eid)
 
 		/* Indoor Access Point must include an additional TPE for
 		 * subordinate devices */
-		if (iconf->he_6ghz_reg_pwr_type == HE_6GHZ_INDOOR_AP) {
+		if (iconf->he_6ghz_reg_pwr_type ==
+		    HE_REG_INFO_6GHZ_AP_TYPE_INDOOR) {
 			/* TODO: Extract PSD limits from channel data */
 			tx_pwr = REG_PSD_MAX_TXPOWER_FOR_SUBORDINATE_CLIENT * 2;
 			eid = hostapd_add_tpe_info(eid, tx_pwr_count,
diff --git a/src/ap/ieee802_11_he.c b/src/ap/ieee802_11_he.c
index bcd547020..c3c54893a 100644
--- a/src/ap/ieee802_11_he.c
+++ b/src/ap/ieee802_11_he.c
@@ -289,12 +289,9 @@ u8 * hostapd_eid_he_operation(struct hostapd_data *hapd, u8 *eid)
 			control |= HE_6GHZ_OPER_INFO_CTRL_DUP_BEACON;
 #endif /* CONFIG_VENDOR_MXL */
 
-		if (hapd->iconf->he_6ghz_reg_pwr_type == 1)
-			control |= HE_6GHZ_STANDARD_POWER_AP <<
-				HE_6GHZ_OPER_INFO_CTRL_REG_INFO_SHIFT;
-		else
-			control |= HE_6GHZ_INDOOR_AP <<
-				HE_6GHZ_OPER_INFO_CTRL_REG_INFO_SHIFT;
+		control |= hapd->iconf->he_6ghz_reg_pwr_type <<
+			HE_6GHZ_OPER_INFO_CTRL_REG_INFO_SHIFT;
+
 		*pos++ = control;
 
 		/* Channel Center Freq Seg0/Seg1 */
diff --git a/src/common/ieee802_11_defs.h b/src/common/ieee802_11_defs.h
index c9fa1f5f3..958864ab5 100644
--- a/src/common/ieee802_11_defs.h
+++ b/src/common/ieee802_11_defs.h
@@ -2598,14 +2598,21 @@ enum ieee80211_he_mcs_support {
 #define HE_OPERATION_BSS_COLOR_MAX		64
 
 /**
- * enum he_6ghz_ap_type - Allowed Access Point types for 6 GHz Band
+ * enum he_reg_info_6ghz_ap_type - Allowed Access Point types for 6 GHz Band
  *
- * IEEE Std 802.11ax-2021, Table E-12 (Regulatory Info subfield encoding in the
- * United States)
+ * IEEE P802.11-REVme/D4.0, Table E-12 (Regulatory Info subfield encoding)
  */
 enum he_6ghz_ap_type {
-	HE_6GHZ_INDOOR_AP		= 0,
-	HE_6GHZ_STANDARD_POWER_AP	= 1,
+	HE_6GHZ_INDOOR_AP               = 0,
+	HE_6GHZ_STANDARD_POWER_AP       = 1,
+};
+enum he_reg_info_6ghz_ap_type {
+	HE_REG_INFO_6GHZ_AP_TYPE_INDOOR         = 0,
+	HE_REG_INFO_6GHZ_AP_TYPE_SP		= 1,
+	HE_REG_INFO_6GHZ_AP_TYPE_VLP		= 2,
+	HE_REG_INFO_6GHZ_AP_TYPE_INDOOR_ENABLED	= 3,
+	HE_REG_INFO_6GHZ_AP_TYPE_INDOOR_SP	= 4,
+	HE_REG_INFO_6GHZ_AP_TYPE_MAX = HE_REG_INFO_6GHZ_AP_TYPE_INDOOR_SP,
 };
 #define HE_OPERATION_BSS_COLOR				0x3f
 
-- 
2.43.0

